espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.java

```diff
@@ -0,0 +1,250 @@
+                w3cActions = convertPress(TAP_TIMEOUT - TIMEOUT_BUFFER);
+                break;
+            case RELEASE:
+                w3cActions = Collections.singletonList(convertRelease());
+                break;
+            case WAIT:
+                w3cActions = Collections.singletonList(convertWait());
+                break;
+            case CANCEL:
+                w3cActions = Collections.singletonList(convertCancel());
+                break;
+            default:
+                throw new InvalidArgumentException(String.format("Unsupported action type %s", action));
+        }
+
+        // All touch actions map to 3 actions
+        return padActionsList(w3cActions);
+    }
+
+    private Action convertCancel() {
+        return new ActionBuilder()
+                .withType(POINTER_CANCEL)
+                .build();
+    }
+
+    private Action convertWait() {
+        return new ActionBuilder()
+                .withType(PAUSE)
+                .build();
+    }
+
+    private Action convertRelease() {
+        return new ActionBuilder()
+                .withType(POINTER_UP)
+                .build();
+    }
+
+    private Origin getOrigin() {
+        Origin origin = new Origin();
+        if (options.getElementId() != null) {
+            origin.setType(InputSource.ELEMENT);
+            origin.setElementId(options.getElementId());
+        } else {
+            origin.setType(InputSource.VIEWPORT);
+        }
+        return origin;
+    }
+
+    private Action getMoveTo() {
+        return new ActionBuilder()
+                .withType(POINTER_MOVE)
+                .withX(options.getX())
+                .withY(options.getY())
+                .withOrigin(getOrigin())
+                .build();
+    }
+
+    private Action convertMoveTo() {
+        return getMoveTo();
+    }
+
+    private List<Action> convertPress(Long pressDuration) {
+        // Move to spot
+        Action moveAction = getMoveTo();
+
+        // Press down
+        Action downAction = new ActionBuilder()
+                .withType(POINTER_DOWN)
+                .build();
+
+        // Wait for the press duration
+        Action waitAction = new ActionBuilder()
+                .withType(PAUSE)
+                .withDuration(pressDuration)
+                .build();
+
+        List<Action> ret = new ArrayList<>();
+        ret.add(moveAction);
+        ret.add(downAction);
+        ret.add(waitAction);
+
+        return ret;
+    }
+
+    private Action getPause() {
+        return new ActionBuilder()
+                .withType(PAUSE)
+                .withDuration(0L)
+                .build();
+    }
+
+    // If an action list has fewer than three actions, pad them with 'pauses' of 0 duration
+    private List<Action> padActionsList(List<Action> actions) {
+        // One Jsonwp Touch Action maps to three W3C Actions
+        final int W3C_ACTIONS_PER_TOUCH_ACTION = 3;
+
+        List<Action> paddedActions = new ArrayList<>(3);
```

Comment:
why this is necessary?

Reply:
For multi-touch actions we need each event to happen synchronously with eachother

So if one input calls press (which maps to `move + down + wait`) and another input is calling pause (which maps to `wait`) we need to add two no-ops to the `wait` event so that it doesn't prematurely advance to the next action.

Reply:
I'd like to see this comment in the code